libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2537,18 +2565,23 @@ void TypeChecker::endVisit(NewExpression const& _newExpression)
 		if (contract->abstract())
 			m_errorReporter.typeError(4614_error, _newExpression.location(), "Cannot instantiate an abstract contract.");
 
-		solAssert(!!m_currentContract, "");
-		m_currentContract->annotation().contractDependencies.insert(contract);
-		solAssert(
-			!contract->annotation().linearizedBaseContracts.empty(),
-			"Linearized base contracts not yet available."
-		);
-		if (contractDependenciesAreCyclic(*m_currentContract))
-			m_errorReporter.typeError(
-				4579_error,
-				_newExpression.location(),
-				"Circular reference for contract creation (cannot create instance of derived or same contract)."
+		if (m_currentContract)
+		{
+			// TODO this is not properly detecting creation-cycles if they go through
+			// internal library functions or free functions. It will be caught at
+			// code generation time, but it would of course be better to catch it here.
```

Comment:
This seems like something worth still doing in this PR... This means ``contractDependencies`` may be incorrect in the presence of free functions? We do export that in the JSON AST, so maybe better not just ignore it...
(Haven't really thought this through, though, so I may be wrong - but seems to me that we need to track contract dependencies for free functions and then propagate them to contracts calling them...)

Reply:
It is already incorrect when using internal library functions, so this PR does not make anything worse.

Reply:
Ah ok... do we have an issue for it then at least?

Reply:
Created it: https://github.com/ethereum/solidity/issues/9565